### DISCUSSION TITLE
string interpolation results in query type never

### DISCUSSION THREAD

**Discussion by zxt-tzx** [0no-co/gql.tada#421] [Q&A] [answered]
### Describe the bug

I'm quite new to this library/GraphQL, so it could well be I'm doing something wrong.

to repro:
- download this GraphQL schema from GitHub [here](https://docs.github.com/en/graphql/overview/public-schema) and put it in my repo
- set up repo/IDE to generate `schema.docs.graphql.d.ts` 
- create the following files
```
import { initGraphQLTada } from "gql.tada";

import type { introspection } from "./schema.docs.graphql.d.ts";

export const graphql = initGraphQLTada<{
  introspection: introspection;
}>();

export type { FragmentOf, ResultOf, VariablesOf } from "gql.tada";
export { readFragment } from "gql.tada";
```

When I try to move my existing query, the type of `const query` is `never`, instead of a `TadaDocument`. 

```TypeScript
import { graphql } from "./graphql";

    const query = graphql(`
    query paginate($cursor: String, $organization: String!, $repo: String!) {
      repository(owner: $organization, name: $repo) {
        issues(
          first: 100
          after: $cursor
          orderBy: { field: UPDATED_AT, direction: ASC }
          filterBy: { since: ${since ? `"${since.toISOString()}"` : "null"} }
        ) {
          nodes {
            id
            number
            title
            body
            url
            state
            stateReason
            createdAt
            updatedAt
            closedAt
            author {
              login
              url
            }
            labels(first: 10) {
              nodes {
                id
                name
                color
                description
              }
            }
            comments(first: 100, orderBy: { field: UPDATED_AT, direction: ASC }) {
              nodes {
                id
                author {
                  login
                  url
                }
                body
                createdAt
                updatedAt
              }
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
        }
      }
      }
    `);
```

I could well be doing something wrong — I am actually just trying to get syntax highlighting + typing when I write the query and I want to pass the query string back in to GitHub's Octokit.

Thanks for the help!

### gql.tada version

"gql.tada": "^1.8.10",

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/0no-co/gql.tada/discussions) should be used
- [X] Read the [docs](https://gql-tada.0no.co/).
- [X] Follow our [Code of Conduct](https://github.com/0no-co/gql.tada/blob/main/CODE_OF_CONDUCT.md)

**Comment by kitten** [comment]
Converted to a discussion since this seems more like a question.

Have you made sure that the editor setup is complete and GraphQLSP is running and giving you editor feedback, e.g. diagnostics?
Most of the time `never` indicates that a type couldn't be inferred. This is either because:
- The typings haven't been exported correctly
- A syntax error is present
The easiest way to resolve the latter is to make sure that GraphQLSP is giving you diagnostics

**Comment by JoviDeCroock** [comment, marked as answer]
You can't interpolate values, you should use document variables. Making the template literal dynamic won't work with TypeScript. https://graphql.org/learn/queries/#variables

**Reply by kitten** [reply]
ah, I didn't see that. Yes, this won't always work correctly in TypeScript inference (but it sometimes can) and won't work with the GraphQLSP.

Just to clarify, even if it sometimes work, you also _shouldn't_ interpolate GraphQL documents.

**Reply by zxt-tzx** [reply]
ah got it. I removed the template literal using TypeScript and the type inference works!

is there a way to convert the resulting `TadaDocument` back to its original query? I know this is a bit of an odd request but it's because I'm using Octokit, which takes in the query as a string rather than a document

thank you very much!

**Reply by JoviDeCroock** [reply]
You can use 'print' from 0no-co/graphql.web or the graphql dependency. A parsed ast can always be printed back to its original text.

**Reply by zxt-tzx** [reply]
cheers, it's working now. I also changed the title so that it's more searchable
